src/aspire/source/xform.py

```diff
@@ -0,0 +1,388 @@
+        def __init__(self, xform, active):
+            self.xform = xform
+            self.active = active
+
+        def __enter__(self):
+            self.xform_old_state = self.xform.active
+            self.xform.active = self.active
+
+        def __exit__(self, exc_type, exc_value, exc_traceback):
+            self.xform.active = self.xform_old_state
+
+    def __init__(self, resolution=np.inf, active=True):
+        """
+        Create a Xform object that works at a specific resolution.
+        :param resolution: The side-length of the Image object on which this Xform is applicable.
+            Note that it is possible to downsample this Xform by later using the `downsample` method on it.
+        :param active: A boolean indicating whether the Xform is active. True by default.
+        """
+        self.resolution = resolution
+        self.active = active
+
+    def forward(self, im, indices=None):
+        """
+        Apply forward transformation for this Xform object to an Image object.
+        :param im: The incoming Image object of depth `n`, on which to apply the forward transformation.
+        :param indices: The indices to use within this Xform. If unspecified, [0..n) is used.
+        :return: An Image object after applying the forward transformation.
+        """
+        if not self.active:
+            return im
+        if indices is None:
+            indices = np.arange(im.n_images)
+        return self._forward(im, indices=indices)
+
+    def _forward(self, im, indices):
+        raise NotImplementedError('Subclasses must implement the _forward method applicable to im/indices.')
+
+    def adjoint(self, im, indices=None):
+        """
+        Apply adjoint transformation for this Xform object to an Image object.
+        :param im: The incoming Image object of depth `n`, on which to apply the adjoint transformation.
+        :param indices: The indices to use within this Xform. If unspecified, [0..n) is used.
+        :return: An Image object after applying the adjoint transformation.
+        """
+        if not self.active:
+            return im
+        if indices is None:
+            indices = np.arange(im.n_images)
+        return self._adjoint(im, indices=indices)
+
+    def _adjoint(self, im, indices):
+        raise NotImplementedError('Subclasses must implement the _adjoint method applicable to im/indices.')
+
+    def enabled(self):
+        """
+        Enable this Xform in a context manager, regardless of it's `active` attribute value.
+        :return: A context manager in which this Xform is enabled.
+        """
+        return Xform.XformActiveContextManager(self, active=True)
+
+    def disabled(self):
+        """
+        Disable this Xform in a context manager, regardless of it's `active` attribute value.
+        :return: A context manager in which this Xform is disabled.
+        """
+        return Xform.XformActiveContextManager(self, active=False)
+
+    def downsample(self, resolution):
```

Comment:
I keep going back and forth about whether this should be in here or not. It's a bit confusing that there's a `downsample` method, but also `Downsample` class. How are they related? This may be indicative of some deeper issue.

Reply:
With there being no model-pipeline anymore, the operation of downsampling a whole Pipeline is now gone. (We just deal with a `Downsample` Xform class).